### ISSUE TITLE
Container logs broken after a checkpoint restore

### ISSUE THREAD

**Issue by CGNonofr** [containers/podman#11677]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When trying the https://criu.org/Podman demo I faced a problem: When you restore a container, the new container stdout is no longer available when doing `podman logs`

**Steps to reproduce the issue:**

1. `podman run -d --name looper busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'`
2. `podman logs -f looper` (a new ligne appears every seconds)
3. `podman container checkpoint looper && podman container restore looper`
4. `podman logs -f looper` (The previous logs are still available but no new ligne appears

**Describe the results you received:**
Nothing is happening in the `podman logs -f looper` output

**Describe the results you expected:**
A new ligne should appears every second with the new counter value

**Additional information you deem important (e.g. issue happens only occasionally):**
The container seems to be alive, because if I replace the write to stdout by a write to a file:
```bash
podman run -d --name looper -v `pwd`:/test busybox /bin/sh -c 'i=0; while true; do echo $i >> /test/counter.txt; i=$(expr $i + 1); sleep 1; done'
tail -f counter.txt
podman container checkpoint looper && podman container restore looper
tail -f counter.txt
```
A new ligne appears in counter.txt every seconds after the restore

I tried in ubuntu 20.04, ubuntu 21.04, with podman 3.0.1 (from the ubuntu repo) and the 3.3.1 with the last criu version

I tried with all the possible configurations (with/without exporting and with every available options)

**Output of `podman version`:**

```
Version:      3.3.1
API Version:  3.3.1
Go Version:   go1.16.6
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/amd64
```

```
# criu --version
Version: 3.15
GitID: v3.14-524-g852d99b35
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.22.3
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - rdma
  cgroupManager: systemd
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/libexec/podman/conmon'
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.0.27, commit: '
  cpus: 12
  distribution:
    distribution: ubuntu
    version: "20.04"
  eventLogger: journald
  hostname: nonofr-MS-7B48
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.4.0-84-generic
  linkmode: dynamic
  memFree: 26781134848
  memTotal: 33613361152
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.20.1.5-925d-dirty
      commit: 0d42f1109fd73548f44b01b3e84d04a279e99d2e
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: true
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: 'slirp4netns: /usr/bin/slirp4netns'
    version: |-
      slirp4netns version 1.1.8
      commit: unknown
      libslirp: 4.3.1-git
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.4.3
  swapFree: 2147479552
  swapTotal: 2147479552
  uptime: 15m 29.94s
registries:
  search:
  - docker.io
  - quay.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 1
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.3.1
  Built: 0
  BuiltTime: Thu Jan  1 01:00:00 1970
  GitCommit: ""
  GoVersion: go1.16.6
  OsArch: linux/amd64
  Version: 3.3.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman/inconnu,now 100:3.3.1-1 amd64  [installé]
podman/inconnu 100:3.3.1-1 arm64
podman/inconnu 100:3.3.1-1 armhf
podman/inconnu 100:3.3.1-1 s390x
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Ubuntu 20.04 physical, Ubuntu 21.04 in VirtualBox

**Comment by vrothberg** [comment]
Thanks for reaching out.  I can reproduce.

It does not seem like an issue in `podman logs`; running `sudo journalctl CONTAINER_NAME=looper` also shows only the logs until the checkpoint / restart.

@adrianreber can you have a look?

**Comment by adrianreber** [comment]
Right. I can also see it. This is probably related to Podman switching to journald logging as default. If I switch to the old logging mechanism (`file`) it still works.

I guess the checkpointing needs to extract the log from journald during checkpointing and insert it back into the journal during restore. The file based logging only added the log file `ctr.log` to the checkpoint archive and put it back during restore. With journald based logging this breaks.

@vrothberg do you think it would be correct to extract the log during checkpointing and insert it back into the journal during restore? Inserting it back into the journal will probably result in wrong time stamps. Maybe it would be useful to have a marker in the journal that the container log is from a restore. And comments or ideas what would be the best here?

**Comment by vrothberg** [comment]
Thank you for checking, @adrianreber.

> @vrothberg do you think it would be correct to extract the log during checkpointing and insert it back into the journal during restore? 

I don't think that's necessary since the previous logs persist in the journal; they're not deleted by Podman.

> And comments or ideas what would be the best here?

To me, it looks like the restored container does not continue logging. The journal only shows the old logs. Is there something special we need to do?

**Comment by adrianreber** [comment]
> Thank you for checking, @adrianreber.
> 
> > @vrothberg do you think it would be correct to extract the log during checkpointing and insert it back into the journal during restore?
> 
> I don't think that's necessary since the previous logs persist in the journal; they're not deleted by Podman.

Yes, I was thinking about moving the container to another host. On the same host it is not necessary.

> > And comments or ideas what would be the best here?
> 
> To me, it looks like the restored container does not continue logging. The journal only shows the old logs. Is there something special we need to do?

I have not yet looked at the code but I am pretty sure the restore code path does not enable journald logging again. I will try to fix that. This will still leave mean that migrated containers will not have their complete log. But maybe that is correct, because the container didn't log anything on that host.

**Comment by vrothberg** [comment]
> Yes, I was thinking about moving the container to another host. On the same host it is not necessary.

My bad, I was too focused on this issue.  That is a very fair point.  I am honestly not sure how to handle that case; I see beauty in both approaches.

@rhatdan @mheon WDYT?

**Comment by CGNonofr** [comment]
> Right. I can also see it. This is probably related to Podman switching to journald logging as default. If I switch to the old logging mechanism (`file`) it still works.

Is there a workaround then? How to switch to the old logging mechanism?

**Comment by adrianreber** [comment]
> > Right. I can also see it. This is probably related to Podman switching to journald logging as default. If I switch to the old logging mechanism (`file`) it still works.
> 
> Is there a workaround then? How to switch to the old logging mechanism?

Please try: `--log-driver k8s-file`. I hope that is the correct one.

**Comment by CGNonofr** [comment]
It doesn't seem to be it, k8s-file seems to be the default value

```
--log-driver string                        Logging driver for the container (default "k8s-file")
```

and using it doesn't fix anything

**Comment by CGNonofr** [comment]
and using `journald` doesn't seem to fix anything either

**Comment by adrianreber** [comment]
What does `podman inspect` say concerning the log file. I get:
```
            "LogConfig": {
                "Type": "k8s-file",
                "Config": null,
                "Path": "/var/lib/containers/storage/overlay-containers/c4add6d80c738f893f43bf34067e95554dd50a3394868895c1dbeae206d1c8e7/userdata/ctr.log",
                "Tag": "",
                "Size": "0B"
            },
```
And the file `ctr.log` contains the file. This is empty if I use `--log-driver journald`.

**Comment by adrianreber** [comment]
Just tested it a bit more and it seems to be a crun bug. If I start the container with `--runtime runc` it just works. So crun is not correctly connecting stdio back on restore. For logging to a file it works. As I wrote the crun CRIU integration this is probably my fault :wink:

**Comment by CGNonofr** [comment]
Ok I confirm `--runtime runc` works! What are the downside of using `runc` instead of `crun`?

**Comment by vrothberg** [comment]
Great catch, @adrianreber! @giuseppe PTAL
